[extensions][py] Gemini

Ankush-lastmile opened the pull request:
[extensions][py] Gemini

[extensions][py] Gemini



## What

Gemini Model Parser Extension

Supports Gemini Models.

Gemini api seems to only allow candidate of 1. This means number of outputs is always 1.

Gemini models are:
Gemini Pro
Gemini pro Visio (similar to gpt -4v)

This model parser supports Gemini Pro T2T functionality


## TestPlan:

Cookbook: https://github.com/lastmile-ai/aiconfig/tree/main/cookbooks/Gemini

## Issue

#463

saqadri commented on extensions/Gemini/python/README.md:
```diff
@@ -0,0 +1,15 @@
+# Using LLama-Guarde for model parsing with AIConfig
```
Update

saqadri commented:
Since PaLM is already part of the the base aiconfig, should we add Gemini directly to aiconfig core? Just a thought, not tied to the idea

Ankush-lastmile commented:
> Since PaLM is already part of the the base aiconfig, should we add Gemini directly to aiconfig core? Just a thought, not tied to the idea

We'll do this in the future, gemini uses the same library as PaLM

Ankush-lastmile commented:
One more type error on colab, removed the type in attempt to get the google-generative ai library working on colab

rholinshead commented on extensions/Gemini/python/README.md:
```diff
@@ -0,0 +1,8 @@
+
+We are prioritizing the Python library Usage
+
+1. Install `pip install aiconfig-extensions-gemini`
```
```suggestion
1. Install `pip install aiconfig-extension-gemini`
```

rholinshead commented on extensions/Gemini/python/README.md:
```diff
@@ -0,0 +1,8 @@
+1. Install `pip install aiconfig-extensions-gemini`
+2. Import the extension `import aiconfig_extension_gemini`
+
+And you're all good to go! This extension automatically updates the model parser registry with the model "Gemini-pro". Check out the Gemini Cookbook in the aiconfig repo for a full example.
```
We should make sure we're consistent with the model name between this and workbooks, for when we download as aiconfig, right? cc @saqadri 

Also, should this be `gemini-pro` (lowercase)?

Ankush-lastmile replied:
This could also be updated in the adapter logic if its not consistent. There are already a couple of differences in aiconfig vs workbooks right now.

saqadri replied:
It should be lowercased here too

rholinshead commented on extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:
```diff
@@ -0,0 +1,295 @@
+                "output_type": "execute_result",
+                "data": candidate.content.parts[0].text,
+                "execution_count": i,
+                "metadata": response.prompt_feedback.__dict__,
```
Where is prompt_feedback defined for Gemini? I've been looking at https://cloud.google.com/vertex-ai/docs/generative-ai/model-reference/gemini#gemini-pro for the req/res but does `genai` sdk add this data?

Ankush-lastmile replied:
The Python sdk Docs were pretty lacking when building this, it was pretty inconsistent throughout their api docs and some of their guides didn't work. Also note that I mainily used their public api docs (not through vertex). Link: https://ai.google.dev/docs

The response object had a couple of attributes but it didn't make sense how to use all of them. Printing the object didn't give me any information either so I inspected it by doing `print(response.__dict__)` which shows me the object fields. Accessing fields like 

<img width="812" alt="Screenshot 2023-12-14 at 5 22 55 PM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/ea85a42a-fefa-453e-966f-c3399f62b77c">

rholinshead commented on extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:
```diff
@@ -0,0 +1,295 @@
+        )
+        await ai_config.callback_manager.run_callbacks(event)
+
+        raise Exception("Not Implemented")
```
```suggestion

```

Should be removed and implemented

saqadri replied:
@Ankush-lastmile make sure this is tracked and you don't close the GH issue until this is done.

Ankush-lastmile replied:
Left a comment on #463. This is my next high-pri focus

rholinshead commented on extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:
```diff
@@ -0,0 +1,295 @@
+            stream = options.stream
+        elif "stream" in completion_data:
+            stream = completion_data["stream"]
+        completion_data["stream"] = stream
```
Is this actually a supported param?

Ankush-lastmile replied:
Yep, there were 4 main attributes: 

<img width="503" alt="Screenshot 2023-12-14 at 5 26 46 PM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/0c3bf913-09ef-463c-a678-c96d88419257">

rholinshead commented on extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:
```diff
@@ -0,0 +1,295 @@
+            if isinstance(output_data, str):
+                return output_data
+            else:
+                raise Exception("Not Implemented")
```
Shouldn't happen, right?

Ankush-lastmile replied:
Not expected, but I hadn't thoroughly understood the expected result types. I put this in as a just in case

jonathanlastmileai replied:
Seems like a `ValueError`

saqadri replied:
Agreed let's update the exception at least

rholinshead commented on extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:
```diff
@@ -0,0 +1,295 @@
+                    )
+
+                    messages.append(
+                        {"role": "user", "parts": [{"text": previous_prompt_template}]}
```
Just double-checking lowercase role works instead of USER and MODEL ?

Ankush-lastmile replied:
Not sure if it matters, the palm equivalent was `author: 0`, `author: 1` but the api/model only checked that it was alternating. Fairly certain that the Gemini is doing something similar.

the lowercase user and model worked for this

rholinshead commented on extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:
```diff
@@ -0,0 +1,295 @@
+
+
+def refine_chat_completion_params(model_settings):
+    # completion parameters to be used for Palm's chat completion api
```
```suggestion
    # completion parameters to be used for Gemini's chat completion api
```

rholinshead commented on extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:
```diff
@@ -0,0 +1,295 @@
+        "contents",
+        "generation_config",
+        "safety_settings",
+        "stream",
```
'tools' can be added later on as well. For generation_config, do we want that to be sub-keyed under model_settings, e.g.
```
model.settings = { generation_config: {topK: ..., etc.} }
```

Or serialized top-level in the settings:
```
model.settings = {topK: ...}
```

The web app is doing the latter I think (cc @saqadri ).

Do we need to match camelCase of https://cloud.google.com/vertex-ai/docs/generative-ai/model-reference/gemini#gemini-pro or does the api here support this snake case?

rholinshead commented:
For the test plan, can you link this local package and test in the cookbook?

jonathanlastmileai commented on extensions/Gemini/python/pyproject.toml:
```diff
@@ -0,0 +1,45 @@
+
+# Black formatting
+[tool.black]
+line-length = 99
```
I don't think we need this section in these toml files if we're using vscode settings?

Ankush-lastmile replied:
can you confirm this is the case? What's the difference otherwise?

jonathanlastmileai commented on extensions/Gemini/python/src/aiconfig_extension_gemini/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .Gemini import GeminiModelParser
+
+__ALL__ = ["GeminiModelParser"]
```
I think this is supposed to be lowercase
```suggestion
__all__ = ["GeminiModelParser"]
```

Ankush-lastmile commented:
Addressed Nits

saqadri commented on extensions/Gemini/python/pyproject.toml:
```diff
@@ -0,0 +1,45 @@
+requires = ["setuptools", "wheel"]
+
+[project]
+name = "aiconfig_extension_gemini"
```
It should have dashes instead of underscores, right? That's how the other extensions are, let's be consistent on names

Ankush-lastmile replied:
the pyproject.toml can not contain dashes, underscores are required here. This will break the package build system. The import statement also uses underscores
ex:
```python
import aiconfig_extension_gemini
```

Where as the pip install needs dashes
```
pip install aiconfig-extension-gemini
```

saqadri replied:
Ah, understood. Thanks for the clarification

Ankush-lastmile commented:
- Updated `get_output_text` to throw ValueError instead of base Exception
- updated init.py `__ALL__` -> `__all__`

rebase to main again (ontop of fix that un-breaks current tests)

saqadri reviewed (approved):
Approving to unblock.
